### DISCUSSION TITLE
How to enable `hot-reload` for `dioxus-desktop`?

### DISCUSSION THREAD

**Discussion by d4h0** [DioxusLabs/dioxus#752] [Q&A] [answered]
Hi,

The `dioxus-desktop` crate has a `hot-reload` feature that is enabled by default.

Is there anything I need to do to get `hot-reload´ working?

Because if I change anything in my RSX, nothing reloads.

Also, I'm running my app via `cargo run`, but that obviously doesn't re-compile my app source code if anything that isn't supported by the RSX `hot-releoad` feature changes.

Is there any other tool/workflow that is recommended for re-compiling and reloading my app on those changes? The CLI only seems to support Wasm apps.

PS: I'm actually interested in `hot-reload` for `dioxus-liveview` (with `axum`), but there doesn't seem to be any `hot-reload` support so far (please correct me if I'm wrong).

My plan is to use `dioxus-desktop` while developing my app to get `hot-reload` support, and then later to switch to `dioxus-liveview`.

Is there anything wrong about this plan?

**Comment by d4h0** [comment]
I'm still interested in the above, but for testing `hot-reload` I just switched to a Wasm app. The `hot-reload` feature in `dioxus` is definitely faster than any other dev-server with `hot-reload` that I've ever tested (including a dev-server I've build myself in Rust) - pretty impressive!

**Comment by d4h0** [comment]
It seems `dioxus_web::hot_reload::init` (or anything else from the `hot_reload` module) is never called. So `dioxus-desktop` doesn't seem to support `hot-reload` at the moment.

---

I thought about how to support `live-reload` in `dioxus-liveview`. Basic support should be relatively easy to implement via the following workflow:

1. The use of [cargo watch](https://crates.io/crates/cargo-watch) (or something similar) for re-compiling and restarting the app (until the CLI can be used for non-Wasm targets)
2. Sending an at startup generated token to WebSocket clients when they connect (e.g., the startup time of the app)
3. On errors, when WebSocket re-connect (e.g., because the app was re-compiled and restarted), if that token is different to the previously received one, reload the whole page
4. Adding a `hot-reload` feature to `dioxus-liveview` that can be activated via a feature of the app (so that this feature can be activated via `cargo watch`, while developing the app)

It would be great to get some feedback about the above (especially, if the approach isn't reasonable).

I'll have a look at the source code of `dioxus-liveview` now, to see if I'm able to add the needed functionality.

**Comment by ealmloff** [comment, marked as answer]
Currently there is no way to enable hot reloading for desktop, liveview, or TUI. The source code is there for desktop with the intention to enable it the future.

With LiveView hot reloading currently requires access to the VirtualDom directly witch only lives on the server, not the client.

I'm adding support for TUI, LiveView, and Desktop without a CLI in #754.

**Reply by d4h0** [reply]
That sounds great, thanks for the update, @Demonthos.

Do you have any idea when this will be ready for testing?

If it is going to be more than a few days – which obviously would be absolutely okay – I might finish the simple implementation I've started today (for my personal use, until official support is ready).

It seems, just re-connecting to the server, when the WebSocket connection is closed, and using something like `cargo watch` is all that's required to get basic hot-reloading working with `liveview`.

**Reply by ealmloff** [reply]
You can test the branch right now, I'm just working on finishing up the documentation.

Just add hot reload init to the top of your main function:
```rust
fn main() {
    hot_reload_init!();
    // launch your application
}
```

The hot reloading this enables will be much faster than cargo watch because it doesn't require a recompile, but it only works on static parts of rsx macro. The approach you mentioned should work for all changes. Eventually we can combine the two approaches with the Dixous CLI to get the best of both worlds.
